api/src/OpenTelemetry/Trace/Core.hs

```diff
@@ -505,6 +499,22 @@ addEvent (FrozenSpan _) _ = pure ()
 addEvent (Dropped _) _ = pure ()
 
 
+-- | Add a link to a recording span.
+addLink :: (MonadIO m) => Span -> NewLink -> m ()
+addLink (Span s) l = liftIO $ do
+  modifyIORef' s $ \(!i) -> i {spanLinks = appendToBoundedCollection (spanLinks i) (freezeLink (spanTracer i) l)}
+addLink (FrozenSpan _) _ = pure ()
+addLink (Dropped _) _ = pure ()
```

Comment:
I see this casing happens a lot throughout this module.

```console
% g grep 'Dropped.*pure ()$'
api/src/OpenTelemetry/Trace/Core.hs:addAttribute (Dropped _) _ _ = pure ()
api/src/OpenTelemetry/Trace/Core.hs:addAttributes (Dropped _) _ = pure ()
api/src/OpenTelemetry/Trace/Core.hs:addEvent (Dropped _) _ = pure ()
api/src/OpenTelemetry/Trace/Core.hs:setStatus (Dropped _) _ = pure ()
api/src/OpenTelemetry/Trace/Core.hs:alterFlags (Dropped _) _ = pure ()
api/src/OpenTelemetry/Trace/Core.hs:updateName (Dropped _) _ = pure ()
api/src/OpenTelemetry/Trace/Core.hs:endSpan (Dropped _) _ = pure ()
api/src/OpenTelemetry/Trace/Core.hs:whenSpanIsRecording (Dropped _) _ = pure ()
```

I think a shared function like,

```hs
withSpan f = \case
  Span s -> f s
  FrozenSpan{} -> pure ()
  Dropped{} -> pure ()
```

Could be nice to have.